### PR TITLE
Fix StringAsEnumResolver and add tests for issue #3524

### PR DESCRIPTION
**Pull request by xuzhg**
Fix StringAsEnumResolver and add tests for issue #3524

- Updated StringAsEnumResolver to handle edge cases in enum resolution
- Added tests in ODataUriResolverTests for StringAsEnumResolver behavior
- Added tests in ODataUriParserTests to verify enum parsing scenarios
- Added UriBuilderTests for round-trip URI building with enum values

### CODE REVIEW

File: test/UnitTests/Microsoft.OData.Core.Tests/UriParser/Metadata/ODataUriResolverTests.cs [hunk 1]
```diff
@@ -88,6 +89,25 @@ private static void VerifyEnumVsStringFilterExpressionReverse(FilterClause filte
...
+        {
+            var uri = new Uri(uriString);
+            var parser = new ODataUriParser(HardCodedTestModel.TestModel, ServiceRoot, uri)
+            {
+                Resolver = new StringAsEnumResolver()
+            };
+
+            Uri rebuiltUri = parser.ParseUri().BuildUri(ODataUrlKeyDelimiter.Parentheses);
+
+            Assert.Equal(WebUtility.UrlDecode(uri.Query), WebUtility.UrlDecode(rebuiltUri.Query));
```
[the remark this answers is not in the document]
Comment (xuzhg) [reply test/UnitTests/Microsoft.OData.Core.Tests/UriParser/Metadata/ODataUriResolverTests.cs]: Use .Uri.UnescapeDataString.

File: test/EndToEndTests/Tests/Core/Microsoft.OData.Core.E2E.Tests/UriBuilderTests/UriBuilderTests.cs [hunk 2]
```diff
@@ -453,15 +454,47 @@ public void AbsoluteUriBuildWithEmptyQueryOptions()
...
+
+        // Act
+        ODataUri odataUri = odataUriParser.ParseUri();
+        var actualUri = odataUri.BuildUri(ODataUrlKeyDelimiter.Slash);
+
+        // Assert
+        Assert.Equal(WebUtility.UrlDecode(queryUri.ToString()), WebUtility.UrlDecode(actualUri.ToString()));
+
+        Assert.Equal(queryUriString, WebUtility.UrlDecode(actualUri.Query));
+    }
```
[the remark this answers is not in the document]
Comment (xuzhg) [reply test/EndToEndTests/Tests/Core/Microsoft.OData.Core.E2E.Tests/UriBuilderTests/UriBuilderTests.cs:479]: Use Uri.UnescapeDataString

File: src/Microsoft.OData.Core/UriParser/Resolver/StringAsEnumResolver.cs [hunk 3]
```diff
@@ -35,26 +35,26 @@ public override void PromoteBinaryOperandTypes(
...
                 {
-                    string text = ((ConstantNode)leftNode).Value as string;
+                    string text = leftConstantNode.Value as string;
                     ODataEnumValue val;
                     IEdmTypeReference typeRef = rightNode.TypeReference;
                     if (TryParseEnum(typeRef.Definition as IEdmEnumType, text, out val))
                     {
-                        leftNode = new ConstantNode(val, text, typeRef);
+                        leftNode = new ConstantNode(val, ODataUriUtils.ConvertToUriLiteral(text, ODataVersion.V4), typeRef);
                         return;
```
[the remark this answers is not in the document]
Comment (xuzhg) [reply src/Microsoft.OData.Core/UriParser/Resolver/StringAsEnumResolver.cs:58]: Fixed.

File: test/UnitTests/Microsoft.OData.Core.Tests/UriParser/ODataUriParserTests.cs [hunk 4]
```diff
@@ -1439,7 +1440,32 @@ public void ParseFullyQualifiedEnumWithResolver()
...
+            }
+
+            // Act
+            ODataUri odataUri = odataUriParser.ParseUri();
+            var actualUri = odataUri.BuildUri(ODataUrlKeyDelimiter.Slash);
+
+            // Assert
+            Assert.Equal(WebUtility.UrlDecode(queryUri.ToString()), WebUtility.UrlDecode(actualUri.ToString()));
+
+            Assert.Equal(queryUriString, WebUtility.UrlDecode(actualUri.Query));
```
[the remark this answers is not in the document]
Comment (xuzhg) [reply test/UnitTests/Microsoft.OData.Core.Tests/UriParser/ODataUriParserTests.cs]: Fixed
